ariac_tutorials/competition_interface.py

```diff
@@ -1063,6 +1179,210 @@ def _kts2_camera_cb(self, msg: AdvancedLogicalCameraImageMsg):
+                    slot_part.part.color = self.color_mapping[color]
+                    slot_part.part.type = self.type_mapping[type]
+                    slot_occupancy_msg.slot_list.append(slot_part)
+
+        return slot_occupancy_msg
+
+    # Helper functions for part detection
+    def colorBound(self, color, bound):
+        if bound == "lower":
+            return np.array([self.HSVcolors[color]["hmin"],
+                             self.HSVcolors[color]["smin"],
+                             self.HSVcolors[color]["vmin"]])
+        # elif bound == "upper":
+        return     np.array([self.HSVcolors[color]["hmax"],
+                             self.HSVcolors[color]["smax"],
+                             self.HSVcolors[color]["vmax"]])
+
+    def load_part_templates(self):
+        self.sensor_template = cv2.imread(
+            path.join(get_package_share_directory("ariac_tutorials"), "tutorial_3_assets", "sensor.png"), cv2.IMREAD_GRAYSCALE)
+        self.regulator_template = cv2.imread(
+            path.join(get_package_share_directory("ariac_tutorials"), "tutorial_3_assets", "regulator.png"), cv2.IMREAD_GRAYSCALE)
+        self.battery_template = cv2.imread(
+            path.join(get_package_share_directory("ariac_tutorials"), "tutorial_3_assets", "battery.png"), cv2.IMREAD_GRAYSCALE)
+        self.pump_template = cv2.imread(
+            path.join(get_package_share_directory("ariac_tutorials"), "tutorial_3_assets", "pump.png"), cv2.IMREAD_GRAYSCALE)
+
+        if (not self.sensor_template.shape[0] > 0) or \
+           (not self.regulator_template.shape[0] > 0) or \
+           (not self.battery_template.shape[0] > 0) or \
+           (not self.pump_template.shape[0] > 0):
+            return False
+        return True
+
+    def draw_slot_grid(self, cv_img):
+        # row 0
+        cv2.circle(cv_img, (5+0*63, 25+0*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+1*63, 25+0*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+2*63, 25+0*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+3*63, 25+0*63), 1, (0, 0, 255), 3)
+        # row 1
+        cv2.circle(cv_img, (5+0*63, 25+1*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+1*63, 25+1*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+2*63, 25+1*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+3*63, 25+1*63), 1, (0, 0, 255), 3)
+        # row 2
+        cv2.circle(cv_img, (5+0*63, 25+2*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+1*63, 25+2*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+2*63, 25+2*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+3*63, 25+2*63), 1, (0, 0, 255), 3)
+        # row 3
+        cv2.circle(cv_img, (5+0*63, 25+3*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+1*63, 25+3*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+2*63, 25+3*63), 1, (0, 0, 255), 3)
+        cv2.circle(cv_img, (5+3*63, 25+3*63), 1, (0, 0, 255), 3)
+
+    def get_tray_ids(self, tray_table_number):
```

Comment:
Are we going to add this to the tutorial?

Reply:
This visualizes the boundaries of the slots. I used it for debugging, should I (1)leave it in there, (2) add it to the visualization on the debug topic or (3) remove it?